### ISSUE TITLE
 Test fails when there is another `tests` folder

### ISSUE THREAD

**Issue by brandsimon** [zopefoundation/zope.testrunner#94]
I have a setup like this:
```
my_module/
my_module/my_file.py
tests/
tests/tests.py
tests/my_file.py
```

tests/tests.py imports the test for my_file and has a `def test_suite():` returning a test suite.
I run the test like this:
`.venv/bin/zope-testrunner --path . -s .tests --color --shuffle`

My problem is, if there is another tests folder somewhere (e.g. `.venv/src/some_github_repo`), zope-testrunner will pick this folder and execute these tests.

I am not sure if this is a testrunner bug or if there is just something wrong with my setup.
If there is information missing, I am happy to provide it.

**Comment by brandsimon** [comment]
Forgot to mention: If I remove the tests folder in .venv/src, everything works as expected

**Comment by d-maurer** [comment]
Simon Brand wrote at 2019-8-1 11:08 -0700:
>I have a setup like this:
>```
>my_module/
>my_module/my_file.py
>tests/
>tests/tests.py
>tests/my_file.py
>```
>
>tests/tests.py imports the test for my_file and has a `def test_suite():` returning a test suite.
>I run the test like this:
>`.venv/bin/zope-testrunner --path . -s .tests --color --shuffle`
>
>My problem is, if there is another tests folder somewhere (e.g. `.venv/src/some_github_repo`), zope-testrunner will pick this folder and execute these tests.
>
>I am not sure if this is a testrunner bug or if there is just something wrong with my setup.
>If there is information missing, I am happy to provide it.

`zope.testrunner` has an option via which you can tell it in which
path is should look for tests.

**Comment by brandsimon** [comment]
@d-maurer 
Thank you, the description of `--test-path` seams what I want, but I am still not able to make it work.

`.venv/bin/zope-testrunner --path . --test-path . -s tests --color --shuffle`
This is runnig the other tests folder (I can understand why)

These commands:
```
.venv/bin/zope-testrunner --path . --test-path tests -s tests --color --shuffle
.venv/bin/zope-testrunner --path . --test-path tests -s . --color --shuffle
```
always result in this error:
```
Traceback (most recent call last):
  File ".venv/bin/zope-testrunner", line 11, in <module>
    load_entry_point('my_module', 'console_scripts', 'zope-testrunner')()
  File ".venv/lib/python3.6/site-packages/zope/testrunner/__init__.py", line 31, in run
    warnings=warnings)
  File ".venv/lib/python3.6/site-packages/zope/testrunner/__init__.py", line 51, in run_internal
    runner.run()
  File ".venv/lib/python3.6/site-packages/zope/testrunner/runner.py", line 179, in run
    feature.global_setup()
  File ".venv/lib/python3.6/site-packages/zope/testrunner/find.py", line 518, in global_setup
    tests = find_tests(self.runner.options, self.runner.found_suites)
  File ".venv/lib/python3.6/site-packages/zope/testrunner/find.py", line 173, in find_tests
    for suite in found_suites:
  File ".venv/lib/python3.6/site-packages/zope/testrunner/find.py", line 194, in find_suites
    for fpath, package in find_test_files(options):
  File ".venv/lib/python3.6/site-packages/zope/testrunner/find.py", line 260, in find_test_files
    for f, package in find_test_files_(options):
  File ".venv/lib/python3.6/site-packages/zope/testrunner/find.py", line 288, in find_test_files_
    for (p, package) in test_dirs(options, {}):
  File ".venv/lib/python3.6/site-packages/zope/testrunner/find.py", line 343, in test_dirs
    p = import_name(p)
  File ".venv/lib/python3.6/site-packages/zope/testrunner/find.py", line 417, in import_name
    __import__(name)
ValueError: Empty module name
```

**Comment by mgedmin** [comment]
Maybe drop the `-s tests` from the command line?  What it does is it tries to `import tests`, and then it recursively searches `tests.__path__[0]` for test files.

Instead I'd try to use --tests-path=tests.

**Comment by mgedmin** [comment]
Oops closed the issue by accidentally clicking the wrong button.

**Comment by mgedmin** [comment]
(Although maybe it should be closed?  zope.testrunner is Working As Designed, even if it is Badly Designed.  I've #93 already open for complaining about the unintuitive behavior of `-s`.)

**Comment by brandsimon** [comment]
@mgedmin 
Then I get the following error:
```
.venv/bin/zope-testrunner --path . --test-path tests --color --shuffle
Test-module import failures:
Module: tests
TypeError: Module tests does not define any tests
Test-modules with import problems:
  tests
```
If this is not a bug, we can close it (but I'd like to have my problem resolved, deleting some tests-directories somewhere cant be a solution)

**Comment by mgedmin** [comment]
Ah, right.  The problem is that you have two entries on sys.path: `.` (added by `--path .`) and `tests` (added by `--test-path tests`).  Then zope.testrunner finds a tests.py and thinks it can be imported by `import tests`, which would work if `.` were not on the path, but since it is, and since this is Python 3 (no `__init__.py` required), then the earlier `.` entry works and you get the `tests` "package" imported.

I don't know what to suggest.  The zope test conventions use the following directory layout:

    setup.py
    src/
        my_package/
            __init__.py
            foo.py
            tests/
                __init__.py
                test_foo.py
            subpackage/
                __init__.py
                bar.py
                tests/
                    __init__.py
                    test_bar.py

and here `zope-testrunner --test-path=src` works fine.

Maybe it would be best to fix the `-s` flag to only consider paths directly inside `--test-path` (i.e. fix #93).  Only I'm not sure how, or I'd've already done it...

Maybe consider a better-maintained test runner like pytest?

**Comment by brandsimon** [comment]
@mgedmin Ok, thank you, I will use another test runner than.
